Documents impersonation (#6293) (#6365)
gravitational/teleport: docs/pages/access-controls/guides/impersonation.mdx

---
title: Impersonating Teleport Users
description: How to issue short-lived certs on behalf of Teleport users using impersonation.
---

# Impersonation

Sometimes users need to create short-lived certificates for non-interactive users, for example CI/CD systems.
Your programs interacting with Teleport may need to create their own authentication as well.
Teleport's impersonation allows users and robots to create short-lived certs for other users and roles.

Let's explore how interactive user Alice can create credentials for a non-interactive
CI/CD user Jenkins and a security scanner.

## Prerequisites

- Installed [Teleport](../getting-started.mdx) or [Teleport Cloud](../../cloud.mdx) >= {{teleport_version}}
- [Tctl admin tool](https://goteleport.com/teleport/download) >= {{teleport_version}}

Verify that your Teleport client is connected:

```bash
$ tctl status
Cluster  acme.example.com
Version  6.0.2
Host CA  never updated
User CA  never updated
Jwt CA   never updated

> CA pin   sha256:e63c7c44be468d37a5b0276b70e9d10b17f24f4be19d6b579810fc94eaa31783

Author: That is not necessary, they can run `tctl` anywhere

> CA pin   sha256:e63c7c44be468d37a5b0276b70e9d10b17f24f4be19d6b579810fc94eaa31783

Reviewer: They can run `tctl status` anywhere, but creating roles (even when logged in as a user with `admin` role) fails when running remotely. I tested this just yesterday

  $ tctl create -f jenkins.yaml
  ```
</Admonition>

## Step 2/3 Create an Impersonator

Let's create a role `impersonator`. Save this role as `impersonator.yaml`:

```yaml
kind: role
version: v3
metadata:
  name: impersonator
spec:
  # SSH options used for user sessions
  options:
    # max_session_ttl defines the TTL (time to live) of SSH certificates
    # issued to the users with this role.
    max_session_ttl: 10h

  # allow section declares a list of resource/verb combinations that are
  # allowed for the users of this role. by default nothing is allowed.
  allow:
    impersonate:
      users: ['jenkins']
      roles: ['jenkins']

  # the deny section uses the identical format as the 'allow' section.
  # the deny rules always override allow rules.
  deny:
    node_labels:
      '*': '*'
```

Create an interactive user `alice` who can access the cluster
and impersonate `jenkins`. We assigned a preset `access` role that allows users to access clusters for
Alice's convenience.

```bash
$ tctl users add alice  --roles=impersonator,access
```

Alice can login using `tsh` and issue a cert for `jenkins`:

```bash
$ tsh login --proxy=teleport.localhost:3080 --user=alice --auth=local

> $ tctl auth sign --user=jenkins --format=openssh --out jenkins --ttl=240h

Reviewer: ```suggestion
$ tctl auth sign --user=jenkins --format=openssh --out=jenkins --ttl=240h
```

  options:
    max_session_ttl: 10h

  # Allow security-impersonator to impersonate any user and role with group: security label
  allow:
    impersonate:
      users: ['*']
      roles: ['*']
      where: 'equals(impersonate_role.metadata.labels["group"], "security") && equals(impersonate_user.metadata.labels["group"], "security")'

  # the deny section uses the identical format as the 'allow' section.
  # the deny rules always override allow rules.
  deny:
    node_labels:
      '*': '*'
```

```bash
$ tctl create -f security-impersonator.yaml
$ tctl users update alice --set-roles=security-impersonator,access
```

Alice can now impersonate any role and user with a label `group: security`.

Create a user and a role `security-scanner`. Save a file `security-scanner.yaml` and create user and group:

```yaml
kind: role
version: v3
metadata:
  name: scanner
  labels:
    group: security
spec:
  options:
    max_session_ttl: 10h

  allow:
    logins: ['root']
    node_labels:
      '*': '*'
---
kind: user
version: v2
metadata:
  name: security-scanner
  labels:
    group: security
spec:
  roles: ['security-scanner']
```

Alice can issue certificates for a security scanner:

``` bash

> $ tctl auth sign --user=security-scanner --format=openssh --out security-scanner --ttl=10h

Reviewer: ```suggestion
$ tctl auth sign --user=security-scanner --format=openssh --out=security-scanner --ttl=10h
```

```

**Matching user traits**

> We can make security-impersonator role more flexible by matching source user and target role parameters:

Reviewer: ```suggestion
We can make the `security-impersonator` role more flexible by matching source user and target role parameters:
```

```yaml
kind: role
version: v3
metadata:
  name: security-impersonator
spec:
  options:
    max_session_ttl: 10h

  allow:
    impersonate:
      users: ['*']
      roles: ['*']
      where: 'contains(user.spec.traits["group"], impersonate_role.metadata.labels["group"]) && contains(user.spec.traits["group"], impersonate_user.metadata.labels["group"])'

  # the deny section uses the identical format as the 'allow' section.
  # the deny rules always override allow rules.
  deny:
    node_labels:
      '*': '*'
```

This role will work with SSO users, but we can test locally, by updating Alice's account
entry with traits:

```yaml
kind: user
version: v2
metadata:
  name: alice
spec:
  traits:
    group: ['security', 'devops']
  roles:
  - security-impersonator
  - access
```

```bash

> # Once Alice logins again, she will receive a new certificate with updated roles.

Reviewer: ```suggestion
# Once Alice logs in again, she will receive a new certificate with updated roles.
```

$ tsh login --proxy=teleport.example.com --user=alice --auth=local

> # Alice can now get a security scanner

Reviewer: ```suggestion
# Alice can now get a certificate for the security scanner
```

> $ tctl --insecure auth sign --user=security-scanner --format=openssh --out security-scanner --ttl=10h

Reviewer: ```suggestion
$ tctl auth sign --user=security-scanner --format=openssh --out=security-scanner --ttl=10h
```

```

> Here is a summary variables and functions we used in this guide:

Reviewer: ```suggestion
Here is a summary of variables and functions we used in this guide:
```
